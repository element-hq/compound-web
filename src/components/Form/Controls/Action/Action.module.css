/*
Copyright 2023 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

.container {
  display: inline-flex;
  position: relative;
}

.control {
  flex: 1;
  padding-inline-end: var(--cpd-space-12x) !important;

  /* From the flexbox spec:
   *   "By default, flex items wonâ€™t shrink below their minimum content size"
   * This allows the element to shrink lower than its natural default size.
   */
  min-inline-size: 0;
}

.action {
  all: unset;
  color: var(--cpd-color-icon-secondary);
  background-color: transparent;
  cursor: pointer;
  position: absolute;
  inset-block: var(--cpd-space-2x);
  inset-inline-end: var(--cpd-space-2x);
  padding: var(--cpd-space-1x);
  overflow: visible;
  border-radius: 50%;
}

.action > svg {
  inline-size: var(--cpd-space-6x);
  block-size: var(--cpd-space-6x);
}

@media (hover) {
  .action:hover {
    color: var(--cpd-color-icon-primary);
    background-color: var(--cpd-color-bg-subtle-secondary);
  }
}

.action:focus-visible {
  outline: 2px solid var(--cpd-color-border-focused);
  outline-offset: 1px;
}

.control[disabled] + .action {
  pointer-events: none;
  color: var(--cpd-color-text-disabled);
}

.control[readonly] + .action {
  pointer-events: none;
  color: var(--cpd-color-text-secondary);
}
