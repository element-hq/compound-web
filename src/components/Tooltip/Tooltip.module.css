/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.tooltip {
  font: var(--cpd-font-body-xs-medium);
  padding: var(--cpd-space-1-5x) var(--cpd-space-3x);
  background: var(--cpd-color-alpha-gray-1400);
  color: var(--cpd-color-text-on-solid-primary);
  border-radius: 4px;
  display: flex;
  flex-direction: column;
  align-items: center;
  text-align: center;
}

.shortcut {
  font-weight: var(--cpd-font-weight-regular);
  color: var(--cpd-color-text-secondary);
}

.arrow {
  /* same color as the tooltip background */
  fill: var(--cpd-color-alpha-gray-1400);
}
