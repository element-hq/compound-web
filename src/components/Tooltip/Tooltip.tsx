/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React, { PropsWithChildren } from "react";
import {
  Root,
  Trigger,
  Portal,
  Content,
  Arrow,
  Provider,
} from "@radix-ui/react-tooltip";

import styles from "./Tooltip.module.css";
import classNames from "classnames";

type TooltipProps = {
  /**
   * The tooltip label
   */
  label: string;
  /**
   * The tooltip caption
   */
  caption?: string;
  /**
   * @deprecated replaced by `caption`
   */
  shortcut?: string;
  /**
   * The side where the tooltip is rendered
   * @default bottom
   */
  side?: React.ComponentProps<typeof Content>["side"];
  /**
   * The preferred alignment against the trigger.
   * May change when collisions occur.
   * @default center
   */
  align?: React.ComponentProps<typeof Content>["align"];
  /**
   * Event handler called when the escape key is down.
   */
  onEscapeKeyDown?: React.ComponentProps<typeof Content>["onEscapeKeyDown"];
  /**
   * Event handler called when a pointer event occurs outside
   * the bounds of the component.
   */
  onPointerDownOutside?: React.ComponentProps<
    typeof Content
  >["onPointerDownOutside"];
  /**
   * The controlled open state of the tooltip.
   * You will mostly want to omit this property. Will be used the vast majority
   * of the time during development.
   * @default undefined
   */
  open?: boolean;
};

/**
 * A tooltip component
 */
export const Tooltip = ({
  children,
  label,
  caption,
  shortcut,
  side = "bottom",
  align = "center",
  onEscapeKeyDown,
  onPointerDownOutside,
  open,
}: PropsWithChildren<TooltipProps>): JSX.Element => {
  return (
    <Provider>
      <Root open={open} delayDuration={300}>
        <Trigger asChild>{children}</Trigger>
        <Portal>
          <Content
            side={side}
            align={align}
            onEscapeKeyDown={onEscapeKeyDown}
            onPointerDownOutside={onPointerDownOutside}
            className={styles.tooltip}
          >
            {label}
            {/* Forcing dark theme, so that we have the correct contrast when
                using the text color secondary on a solid dark background. 
                This is temporary and should only remain until we figure out 
                the approach to on-solid tokens */}
            {(caption || shortcut) && (
              <span className={classNames(styles.caption, "cpd-theme-dark")}>
                {caption ?? shortcut}
              </span>
            )}
            <Arrow width={10} height={6} className={styles.arrow} />
          </Content>
        </Portal>
      </Root>
    </Provider>
  );
};
