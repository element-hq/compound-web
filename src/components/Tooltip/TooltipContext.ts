/*
 * Copyright 2024 New Vector Ltd.
 *
 * SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
 * Please see LICENSE files in the repository root for full details.
 */

import { createContext, useContext } from "react";
import { useTooltip } from "./useTooltip";

type ContextType = ReturnType<typeof useTooltip> | null;
/**
 * The context for the Tooltip components.
 */
export const TooltipContext = createContext<ContextType>(null);

/**
 * Provides the context for the Tooltip components.
 */
export function useTooltipContext() {
  const context = useContext(TooltipContext);

  if (context == null) {
    throw new Error("Tooltip components must be wrapped in <Tooltip />");
  }

  return context;
}
