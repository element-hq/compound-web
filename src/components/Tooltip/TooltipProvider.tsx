import { FloatingDelayGroup } from "@floating-ui/react";
import React, { type FC } from "react";

export const hoverDelay = { open: 300, close: 0 };

interface Props {
  children: React.ReactNode;
}

/**
 * Enables tooltips to share a global "warm-up" period for their hover delay.
 * You must wrap your application in this component for tooltips to function.
 */
export const TooltipProvider: FC<Props> = ({ children }) => (
  <FloatingDelayGroup delay={hoverDelay} timeoutMs={300}>
    {children}
  </FloatingDelayGroup>
);

TooltipProvider.displayName = "TooltipProvider";
