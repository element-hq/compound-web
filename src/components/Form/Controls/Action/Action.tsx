/*
Copyright 2023 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import classnames from "classnames";
import React, { forwardRef, ComponentRef, ComponentProps, useId } from "react";
import styles from "./Action.module.css";
import { TextInput } from "../Text";

import { Control } from "@radix-ui/react-form";
import { Tooltip } from "../../../Tooltip/Tooltip";

type Props = {
  /**
   * The CSS class name.
   */
  className?: string;
  /**
   * The React component to display on the right hand side of the icon.
   */
  Icon: React.FunctionComponent<React.SVGAttributes<SVGElement>>;
  /**
   * On click callback for the icon on the right hand side
   */
  onActionClick: (e: React.MouseEvent) => void;
  /**
   * The name of the action to perform (e.g. "Copy to clipboard")
   */
  actionLabel: string;
} & React.ComponentProps<typeof TextInput>;

/**
 * A generic input with an action icon on the right hand side.
 * You should rarely use it directly,
 * but it is powering the likes of "copy to clipboard input", "password reveal", ...
 */
export const ActionInput = forwardRef<HTMLInputElement, Props>(
  function ActionControl(
    { Icon, className, actionLabel, onActionClick, ...props },
    ref,
  ) {
    const id = useId();
    const classes = classnames(styles.container, className);
    return (
      <div className={classes} id={id}>
        <TextInput ref={ref} {...props} className={styles.control} />

        <Tooltip label={actionLabel}>
          <button
            type="button"
            className={styles.action}
            onClick={onActionClick}
            aria-controls={id}
          >
            <Icon aria-hidden />
          </button>
        </Tooltip>
      </div>
    );
  },
);

/**
 * A styled text input wrapped in a `Control` component, for use in Radix forms.
 */
export const ActionControl = forwardRef<
  ComponentRef<typeof ActionInput>,
  ComponentProps<typeof ActionInput>
>(function ActionControl(props, ref) {
  return (
    <Control asChild>
      <ActionInput ref={ref} {...props} />
    </Control>
  );
});
