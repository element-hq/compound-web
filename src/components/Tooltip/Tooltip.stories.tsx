/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from "react";
import { Meta, StoryFn } from "@storybook/react";

import { Tooltip as TooltipComponent } from "./Tooltip";
import { IconButton } from "../Button";

import UserIcon from "@vector-im/compound-design-tokens/icons/user-profile.svg";

export default {
  title: "Tooltip",
  component: TooltipComponent,
  tags: ["autodocs"],
  argTypes: {},
  args: {},
} as Meta<typeof TooltipComponent>;

const TemplateSide: StoryFn<typeof TooltipComponent> = () => (
  <div
    style={{
      display: "flex",
      gap: "50px",
      flexDirection: "column",
      alignItems: "center",
    }}
  >
    <TooltipComponent open={true} side="top" label="@bob:example.org">
      <IconButton>
        <UserIcon />
      </IconButton>
    </TooltipComponent>
    <TooltipComponent open={true} side="right" label="@bob:example.org">
      <IconButton>
        <UserIcon />
      </IconButton>
    </TooltipComponent>
    <TooltipComponent open={true} side="bottom" label="@bob:example.org">
      <IconButton>
        <UserIcon />
      </IconButton>
    </TooltipComponent>
    <TooltipComponent open={true} side="left" label="@bob:example.org">
      <IconButton>
        <UserIcon />
      </IconButton>
    </TooltipComponent>
  </div>
);

export const Side = TemplateSide.bind({});
Side.args = {};

const TemplateAlign: StoryFn<typeof TooltipComponent> = () => (
  <div
    style={{
      display: "flex",
      gap: "50px",
      flexDirection: "column",
      alignItems: "center",
    }}
  >
    <TooltipComponent open={true} align="center" label="Copy" caption="⌘ + C">
      <IconButton>
        <UserIcon />
      </IconButton>
    </TooltipComponent>
    <TooltipComponent
      open={true}
      align="start"
      label="@bob:example.org"
      caption="⌘ + C"
    >
      <IconButton>
        <UserIcon />
      </IconButton>
    </TooltipComponent>
    <TooltipComponent
      open={true}
      align="end"
      label="@bob:example.org"
      caption="⌘ + C"
    >
      <IconButton>
        <UserIcon />
      </IconButton>
    </TooltipComponent>
  </div>
);

export const Align = TemplateAlign.bind({});
Align.args = {};
