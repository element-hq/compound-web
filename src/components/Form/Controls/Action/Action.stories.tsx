/*
Copyright 2025 New Vector Ltd.
Copyright 2023 The Matrix.org Foundation C.I.C.
Copyright 2023 New Vector Ltd

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import React from "react";

import { Meta, StoryObj } from "@storybook/react-vite";
import { fn } from "storybook/test";

import * as icons from "@vector-im/compound-design-tokens/assets/web/icons";

import { ActionInput } from "./";
import { TooltipProvider } from "../../../Tooltip/TooltipProvider";

type Props = { invalid?: boolean } & React.ComponentProps<typeof ActionInput>;

const meta = {
  title: "Form/Controls/Action",
  component: ActionInput,
  tags: ["autodocs"],
  parameters: {
    controls: {
      include: [
        "onActionClick",
        "defaultValue",
        "placeholder",
        "disabled",
        "autoFocus",
        "readOnly",
        "invalid",
        "actionLabel",
        "Icon",
      ],
    },
  },
  argTypes: {
    onActionClick: {
      action: "clicked",
    },
    defaultValue: {
      type: "string",
    },
    placeholder: {
      type: "string",
    },
    disabled: {
      type: "boolean",
    },
    autoFocus: {
      type: "boolean",
    },
    readOnly: {
      type: "boolean",
    },
    invalid: {
      type: "boolean",
    },
    actionLabel: {
      type: "string",
    },
    Icon: {
      options: Object.keys(icons),
      mapping: icons,
      defaultValue: "check",
      control: {
        type: "select",
      },
    },
  },
  render: ({ invalid, ...restArgs }) => (
    <TooltipProvider>
      <ActionInput data-invalid={invalid || undefined} {...restArgs} />
    </TooltipProvider>
  ),
  args: {
    placeholder: "",
    autoFocus: false,
    disabled: false,
    readOnly: false,
    invalid: false,
    actionLabel: "Action",
    onActionClick: fn(),
    Icon: icons.CheckIcon,
  },
} satisfies Meta<Props>;
export default meta;

type Story = StoryObj<typeof meta>;

export const Empty: Story = {
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/file/rTaQE2nIUSLav4Tg3nozq7/Compound-Web-Components?type=design&node-id=793-4477",
    },
  },
};

export const Filled: Story = {
  args: {
    defaultValue: "Filled",
  },
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/file/rTaQE2nIUSLav4Tg3nozq7/Compound-Web-Components?type=design&node-id=793-4536",
    },
  },
};

export const Disabled: Story = {
  args: {
    defaultValue: "Disabled",
    disabled: true,
  },
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/file/rTaQE2nIUSLav4Tg3nozq7/Compound-Web-Components?type=design&node-id=793-4530",
    },
  },
};

export const ReadOnly: Story = {
  args: {
    defaultValue: "Read only",
    readOnly: true,
  },
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/file/rTaQE2nIUSLav4Tg3nozq7/Compound-Web-Components?type=design&node-id=2126-21984",
    },
  },
};

export const Focused: Story = {
  args: {
    autoFocus: true,
  },
};

export const Invalid: Story = {
  args: {
    defaultValue: "Invalid",
    invalid: true,
  },
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/file/rTaQE2nIUSLav4Tg3nozq7/Compound-Web-Components?type=design&node-id=793-4514",
    },
  },
};
