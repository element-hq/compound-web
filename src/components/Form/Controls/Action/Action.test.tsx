/*
Copyright 2023 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import { vi, describe, it, expect } from "vitest";
import { render, screen } from "@testing-library/react";
import React from "react";
import ChatIcon from "@vector-im/compound-design-tokens/assets/web/icons/chat";

import { ActionInput } from "./Action";
import { TooltipProvider } from "../../../Tooltip/TooltipProvider";
import userEvent from "@testing-library/user-event";

describe("ActionInput", () => {
  it("renders", () => {
    const { asFragment } = render(
      <TooltipProvider>
        <ActionInput
          Icon={ChatIcon}
          actionLabel="Click me!"
          onActionClick={() => {
            console.log("clicked!");
          }}
        />
      </TooltipProvider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("trigger the action", async () => {
    const user = userEvent.setup();
    const spy = vi.fn();

    render(
      <TooltipProvider>
        <ActionInput
          Icon={ChatIcon}
          actionLabel="Click me!"
          onActionClick={spy}
        />
      </TooltipProvider>,
    );

    await user.click(screen.getByRole("button", { name: "Click me!" }));

    expect(spy).toHaveBeenCalledTimes(1);
  });
});
