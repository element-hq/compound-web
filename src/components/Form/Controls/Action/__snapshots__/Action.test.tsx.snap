// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ActionInput > renders 1`] = `
<DocumentFragment>
  <div
    class="_container_1689d3"
    id="«r0»"
  >
    <input
      class="_control_a839aa _control_1689d3"
    />
    <button
      aria-controls="«r0»"
      aria-labelledby="«r1»"
      class="_action_1689d3"
      type="button"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m1.5 21.25 1.45-4.95a10.2 10.2 0 0 1-.712-2.1A10.2 10.2 0 0 1 2 12q0-2.075.788-3.9a10.1 10.1 0 0 1 2.137-3.175q1.35-1.35 3.175-2.137A9.7 9.7 0 0 1 12 2q2.075 0 3.9.788a10.1 10.1 0 0 1 3.175 2.137q1.35 1.35 2.137 3.175A9.7 9.7 0 0 1 22 12a9.7 9.7 0 0 1-.788 3.9 10.1 10.1 0 0 1-2.137 3.175q-1.35 1.35-3.175 2.137A9.7 9.7 0 0 1 12 22q-1.125 0-2.2-.238a10.2 10.2 0 0 1-2.1-.712L2.75 22.5a.94.94 0 0 1-1-.25.94.94 0 0 1-.25-1m2.45-1.2 3.2-.95a1 1 0 0 1 .275-.062q.15-.013.275-.013.225 0 .438.038.212.036.412.137a7.4 7.4 0 0 0 1.675.6Q11.1 20 12 20q3.35 0 5.675-2.325T20 12t-2.325-5.675T12 4 6.325 6.325 4 12q0 .9.2 1.775t.6 1.675q.176.325.188.688t-.088.712z"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
